File: lib/matching/routing.ts
import type { DecisionOutcome } from "@/lib/types";
import { approverRoleForAmount, type PolicyConfig } from "@/lib/matching/policy";

/**
 * CLAUDE.md section 3: an exception never routes to a cheaper approver than
 * a clean invoice of the same amount would get — it routes to that same
 * band PLUS AP. Duplicate holds get no approval task at all. Blocked
 * invoices bypass the amount bands entirely — bank-detail and identity
 * failures escalate straight to AP + Controller regardless of amount.
 */
export function computeApprovalRoute(outcome: DecisionOutcome, totalUsd: number, policy: PolicyConfig): string[] {
  switch (outcome) {
    case "ready_for_approval":
      return [approverRoleForAmount(policy, totalUsd)];
    case "exception_review":
      return [approverRoleForAmount(policy, totalUsd), "ap_review_team"];
    case "duplicate_hold":
      return [];
    case "blocked":
      return ["ap_review_team", "controller"];
  }
}

const COST_CENTER_RULES: Array<{ keywords: RegExp; costCenter: string; accountCode: string }> = [
  { keywords: /clean|janitor|pest/i, costCenter: "CC-FAC-CLEAN", accountCode: "6120-SUPPLIES" },
  { keywords: /hvac|mechanical|compressor|refrigerant/i, costCenter: "CC-FAC-MECH", accountCode: "6130-MAINTENANCE" },
  { keywords: /security|patrol|guard/i, costCenter: "CC-FAC-SEC", accountCode: "6140-SECURITY" },
  { keywords: /ground|landscap|mow|mulch|irrigation/i, costCenter: "CC-FAC-GRND", accountCode: "6150-GROUNDS" },
];

/**
 * Lightweight keyword heuristic — there's no supplier→cost-center mapping in
 * the schema yet, so this infers from supplier name and line descriptions.
 * A known Phase 4 simplification; real cost-center assignment is Phase 5
 * territory once the accounting integration exists to validate against.
 */
export function guessCostCenter(supplierName: string, lineDescriptions: string[]): { costCenter: string; accountCode: string } {
  const haystack = [supplierName, ...lineDescriptions].join(" ");
  const rule = COST_CENTER_RULES.find((r) => r.keywords.test(haystack));
  return rule ? { costCenter: rule.costCenter, accountCode: rule.accountCode } : { costCenter: "CC-FAC-CLEAN", accountCode: "6120-SUPPLIES" };
}

File: ledgerguard/routing.py
"""Approval routing + cost-center guessing."""

import re

from .policy import PolicyConfig, approver_role_for_amount

COST_CENTER_RULES: list[dict] = [
    {"keywords": re.compile(r"clean|janitor|pest", re.I), "cost_center": "CC-FAC-CLEAN", "account_code": "6120-SUPPLIES"},
    {
        "keywords": re.compile(r"hvac|mechanical|compressor|refrigerant", re.I),
        "cost_center": "CC-FAC-MECH",
        "account_code": "6130-MAINTENANCE",
    },
    {"keywords": re.compile(r"security|patrol|guard", re.I), "cost_center": "CC-FAC-SEC", "account_code": "6140-SECURITY"},
    {
        "keywords": re.compile(r"ground|landscap|mow|mulch|irrigation", re.I),
        "cost_center": "CC-FAC-GRND",
        "account_code": "6150-GROUNDS",
    },
]


def compute_approval_route(outcome: str, total_usd: float, policy: PolicyConfig) -> list[str]:
    """An exception never routes to a cheaper approver than a clean invoice of
    the same amount would get — it routes to that same band PLUS AP.
    Duplicate holds get no approval task at all. Blocked invoices bypass the
    amount bands entirely — bank-detail and identity failures escalate
    straight to AP + Controller regardless of amount."""
    if outcome == "ready_for_approval":
        return [approver_role_for_amount(policy, total_usd)]
    if outcome == "exception_review":
        return [approver_role_for_amount(policy, total_usd), "ap_review_team"]
    if outcome == "duplicate_hold":
        return []
    if outcome == "blocked":
        return ["ap_review_team", "controller"]
    return []


def guess_cost_center(supplier_name: str, line_descriptions: list[str]) -> dict:
    """Lightweight keyword heuristic — there's no supplier-to-cost-center
    mapping in the schema, so this infers from supplier name and line
    descriptions."""
    haystack = " ".join([supplier_name, *line_descriptions])
    for rule in COST_CENTER_RULES:
        if rule["keywords"].search(haystack):
            return {"cost_center": rule["cost_center"], "account_code": rule["account_code"]}
    return {"cost_center": "CC-FAC-CLEAN", "account_code": "6120-SUPPLIES"}
